File: app/src/main/java/com/dsh/txlessons/threadandroid/MainActivity.java
package com.dsh.txlessons.threadandroid;

import android.os.AsyncTask;
import android.os.Handler;
import android.os.HandlerThread;
import android.os.Looper;
import android.os.Message;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import com.dsh.mydemos.R;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main2);

        Looper.myLooper();
        Looper.getMainLooper();

        HandlerThread handlerThread = new HandlerThread("Name");
        //handler创建方式1
        Handler handler = new Handler(Looper.getMainLooper(), new Handler.Callback() {
            @Override
            public boolean handleMessage(Message msg) {
                //自定义callBack处理消息
                return false;
            }
        });
        //handler创建方式2
        Handler handler2 = new Handler(handlerThread.getLooper());

        MyAsyncTask task = new MyAsyncTask();
        task.execute("Hello AsyncTask");

    }

    class MyAsyncTask extends AsyncTask{

        @Override protected void onPreExecute() {
            super.onPreExecute();
        }

        @Override protected Object doInBackground(Object[] objects) {
            System.out.println(objects[0]);
            return null;
        }

        @Override protected void onPostExecute(Object o) {
            super.onPostExecute(o);
        }
    }
}

File: app/src/main/java/com/dsh/txlessons/threadrxjava/MainActivity.kt
package com.dsh.txlessons.threadrxjava;

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.widget.TextView
import com.dsh.mydemos.R
import com.dsh.txlessons.threadrxjava.network.Api
import com.dsh.threadrxjava.model.Repo
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.Single
import io.reactivex.SingleObserver
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Function
import io.reactivex.schedulers.Schedulers
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeUnit.SECONDS

class MainActivity : AppCompatActivity() {
  private var disposable: Disposable? = null
  private lateinit var textView:TextView;

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_rxjava)

    textView = findViewById(R.id.textView);

    //rxjava网络请求示例
    rxJavaNetRequest();

    //Single使用
    single();

    //操作符
    map();
    interval();
    delay();


    Single.just(1)
      .subscribeOn(Schedulers.io())
      .observeOn(AndroidSchedulers.mainThread())



  }

  private fun delay() {
    Single.just("1").delay(1,SECONDS);
  }

  private fun interval() {
    Observable.interval(0, 1, TimeUnit.SECONDS)
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(object : Observer<Long?> {
          override fun onComplete() {
          }

          override fun onSubscribe(d: Disposable) {
          }

          override fun onNext(t: Long) {
            textView.text = t.toString()
          }

          override fun onError(e: Throwable) {
          }
        })
  }

  private fun map() {
    var singleInt:Single<Int> = Single.just(1)
    val singleMap:Single<String> = singleInt.map(object :Function<Int,String>{
      override fun apply(t: Int): String {
        return t.toString();
      }
    })
    singleMap.subscribe(object :SingleObserver<String>{
      override fun onSuccess(t: String) {
        textView.text = t;
      }
      override fun onSubscribe(d: Disposable) {}
      override fun onError(e: Throwable) {}
    })
  }

  private fun rxJavaNetRequest() {
    val retrofit = Retrofit.Builder()
        .baseUrl("https://api.github.com/")
        .addConverterFactory(GsonConverterFactory.create())
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())//RXJava支持
        .build()

    val api = retrofit.create(Api::class.java)


    api.getRepos("willA")
        .subscribeOn(Schedulers.io())
        .observeOn(AndroidSchedulers.mainThread())
        .subscribe(object : SingleObserver<List<Repo?>?> {
          override fun onSubscribe(disposable: Disposable) {
            textView.text = "正在请求"
            this@MainActivity.disposable = disposable
          }

          override fun onSuccess(repos: List<Repo?>) {
            textView.text = repos[0]?.name
          }

          override fun onError(e: Throwable) {
            textView.text = e.message ?: e.javaClass.name
          }


        });
  }

  private fun single() {
    var single = Single.just("1")
    single.subscribe(object :SingleObserver<String>{
      override fun onSuccess(t: String) {
        textView.text = t;
      }

      override fun onSubscribe(d: Disposable) {}

      override fun onError(e: Throwable) {}

    })
  }




  override fun onDestroy() {
    disposable?.dispose()
    super.onDestroy()
  }

}





